### PR TITLE
[ERE-2049] Working Group Types and Rules

### PR DESCRIPTION
**Pull request by ppettitau**
* Add new models Working Group Type and Working Group Type Rule
* Modify Working Group multi select in Patient, User, and Report Designer forms to use optgroups to provide visual separation where working groups have types.
* Add logic to use newly defined Working Group Type Rule to determine which working groups a user has default access to, under certain circumstances.

### CODE REVIEW

**Comment by ppettitau** [comment]
Need to do:
- [x] Unit test for `working_group_optgroup_choices` in `rdrf/registry/groups/forms.py`
- [x] Unit test for `WorkingGroupManager.get_by_user()` in `rdrf/registry/groups/models.py`
- [-] Potentially rename `WorkingGroupManager.get_by_user()` to something more descriptive of its purpose i.e. `get_by_user_with_type_rules` (EDIT 12/5 - Decided to introduce a parameter to allow application of type rules to be configurable)

File: rdrf/rdrf/testing/unit/groups_form_tests.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        WorkingGroup.objects.create(id=5, name='Kidney condition B', registry=self.reg_arrk, type=self.type_scg)
+        WorkingGroup.objects.create(id=6, name='Sunshine Clinic', registry=self.reg_other, type=self.type_hospital)
+        WorkingGroup.objects.create(id=7, name='Unallocated', registry=self.reg_arrk)
+        WorkingGroup.objects.create(id=8, name='Unallocated', registry=self.reg_other)
+        WorkingGroup.objects.create(id=9, name='Teaching', registry=self.reg_arrk)
+
+        def custom_make_option_fn(working_group):
+            return json.dumps({'id': working_group.id, 'registry': working_group.registry.code}), f'{working_group.name}'
+
+        choices = working_group_optgroup_choices(WorkingGroup.objects.all(), custom_make_option_fn)
```
Comment (ScriptSmith) [inline rdrf/rdrf/testing/unit/groups_form_tests.py:73]: Not sure what this is testing. Why does it need a custom fn param, and why test multiple registries?

File: rdrf/rdrf/testing/unit/groups_model_tests.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
+from django.contrib.auth.models import Group
+from django.test import TestCase
+
+from registry.groups.models import CustomUser, WorkingGroupType, WorkingGroup, WorkingGroupTypeRule
+
+
+class GroupModelTest(TestCase):
+
+    def setUp(self) -> None:
```
Comment (ScriptSmith) [inline rdrf/rdrf/testing/unit/groups_model_tests.py]: Return type not being checked so not worth including.

File: rdrf/rdrf/testing/unit/groups_model_tests.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+                                                                           self.wg_cardiology,
+                                                                           self.wg_greenslopes,
+                                                                           self.wg_mater,
+                                                                           self.wg_neurology,
+                                                                           self.wg_outpatients,
+                                                                           self.wg_royal,
+                                                                           self.wg_stroke])
+
+        # Turn off apply_type_rules
+        self.assertEqual(list(WorkingGroup.objects.get_by_user(self.user_researcher, apply_type_rules=False)), [self.wg_arrythmia,
```
Comment (ScriptSmith) [inline rdrf/rdrf/testing/unit/groups_model_tests.py]: The `apply_type_rules` param isn't really useful if it's just used in this assertion. Maybe a feature to test this if it's necessary?

**Review by ScriptSmith — CHANGES_REQUESTED** [review]
From the ticket:

> We can also take this further and add rules for the patients, and then use that to validate their groups on save
> e.g. Patients MUST have a Hospital WG and MUST have a Specific Condition Group WG

Was this considered? Seems like it will be needed eventually.

File: rdrf/rdrf/testing/unit/groups_form_tests.py (hunk 1 above)
Comment (ppettitau) [reply rdrf/rdrf/testing/unit/groups_form_tests.py:73]: It is emulating how the function is used for the Report Designer. 

The custom fn param is required to return a json (similar to the test) to the UI which is then used by the JS to hide/show working groups based on the Registry selected.

File: rdrf/rdrf/testing/unit/groups_model_tests.py (hunk 3 above)
Comment (ppettitau) [reply rdrf/rdrf/testing/unit/groups_model_tests.py]: I guess it's not currently necessary, I was just trying to think ahead where we could potentially do some refactoring to use this method. I'll remove it for now

File: rdrf/rdrf/testing/unit/groups_model_tests.py (hunk 2 above)
Comment (ppettitau) [reply rdrf/rdrf/testing/unit/groups_model_tests.py]: Oops. it was autogenerated and I didn't notice it.

Fixed.

File: rdrf/rdrf/testing/unit/groups_model_tests.py (hunk 3 above)
Comment (ppettitau) [reply rdrf/rdrf/testing/unit/groups_model_tests.py]: Fixed.

**Comment by ppettitau** [comment]
> From the ticket:
> 
> > We can also take this further and add rules for the patients, and then use that to validate their groups on save
> > e.g. Patients MUST have a Hospital WG and MUST have a Specific Condition Group WG
> 
> Was this considered? Seems like it will be needed eventually.

Once I got into it, it didn't really feel like it fit in with this feature which is more about assigning default access rather than enforcing requiredness.

File: rdrf/rdrf/testing/unit/groups_form_tests.py (hunk 1 above)
Comment (ScriptSmith) [reply rdrf/rdrf/testing/unit/groups_form_tests.py:73]: Ah yep, my bad! Missed that :facepalm:.

**Review by ScriptSmith — APPROVED** [review]

**Pull request merged**